framework/src/org/checkerframework/common/jaifinference/JaifInferenceUtils.java

```diff
@@ -0,0 +1,485 @@
+     * @param methodElt is the method element.
+     * @param atf is the annotated type factory.
+     * @return the ATM of methodElt in the .jaif file of the class with symbol
+     * classSymbol, or null if it can't be found.
+     */
+    public static AnnotatedTypeMirror getMethodReturnTypeInJaif(
+            ClassSymbol classSymbol, ExecutableElement methodElt,
+            AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return null; // Static block.
+        String jaifPath = JAIF_FILES_PATH + classSymbol.flatname.toString().
+                replaceAll("\\.", "/") + ".jaif";
+        try {
+            AClass clazz = getJaifClass(classSymbol, jaifPath, new AScene());
+            if (clazz == null) return null;
+            String methodName = JVMNames.getJVMMethodName(methodElt);
+            AMethod method = clazz.methods.get(methodName);
+            if (method != null && method.returnType != null) {
+                Set<Annotation> prevAnnos = method.returnType.tlAnnotationsHere;
+                return setOfAnnotationsToATM(prevAnnos, atf, false,
+                            methodElt.getReturnType());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    /**
+     * @param classSymbol is the symbol representing the class containing the method.
+     * @param fieldNode is the node representing the field.
+     * @param atf is the annotated type factory.
+     * @return the ATM of fieldNode in the .jaif file of the class with symbol
+     * classSymbol, or null if it can't be found.
+     */
+    public static AnnotatedTypeMirror getFieldTypeInJaif(
+            ClassSymbol classSymbol, FieldAccessNode fieldNode,
+            AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return null; // Static block.
+        String jaifPath = JAIF_FILES_PATH + classSymbol.flatname.toString().
+                replaceAll("\\.", "/") + ".jaif";
+
+        AScene scene = new AScene();
+        try {
+            AClass clazz = getJaifClass(classSymbol, jaifPath, scene);
+            if (clazz == null) return null;
+            String fieldName = TreeUtils.getFieldName(fieldNode.getTree());
+            AField field = clazz.fields.get(fieldName);
+            if (field == null) {
+                return null;
+            }
+            Set<Annotation> annos = field.tlAnnotationsHere;
+            if (annos != null && annos.size() > 0) {
+                return setOfAnnotationsToATM(annos, atf, false, fieldNode.getType());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    /**
+     * Updates the type of the field lhs in the .jaif file of the class with
+     * symbol classSymbol. The new type will be the LUB between the previous
+     * type and the type of lhs. If the new type is not a subtype of the type of
+     * lhs, no refinement is made.
```

Comment:
I had to read this documentation a few times, because the second sentence says what the new type is, and then the third sentence says that the new type doesn't count sometimes depending on what it is.
Here are two alternate suggestions:
 * Change the documentation to say initially, not at the end, under what conditions no update happens, or
 * Change the specification so that the update is performed unconditionally.  (Maybe the .jaif writing process would filter some of the types.)